worker/multiwatcher/worker.go

```diff
@@ -0,0 +1,352 @@
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &Worker{
+		config: config,
+		// There always needs to be a valid request channel.
+		request: make(chan *request),
+		waiting: make(map[*Watcher]*request),
+		store:   multiwatcher.NewStore(config.Logger),
+	}
+	w.metrics = NewMetricsCollector(w)
+	w.tomb.Go(w.loop)
+	return w, nil
+}
+
+const (
+	// Keys used in the Report method are used to retrieve from
+	// the map in the metrics code, so define constants for the keys.
+	reportWatcherKey = "num-watchers"
+	reportStoreKey   = "store-size"
+	reportRestartKey = "restart-count"
+	reportErrorsKey  = "errors"
+)
+
+// Report is shown up in the engine report of the agent.
+func (w *Worker) Report() map[string]interface{} {
+	w.mu.Lock()
+	count := len(w.watchers)
+	store := w.store.Size()
+	restart := w.restartCount
+	var errors []string
+	for _, err := range w.errors {
+		errors = append(errors, err.Error())
+	}
+	w.mu.Unlock()
+
+	report := map[string]interface{}{
+		reportWatcherKey: count,
+		reportStoreKey:   store,
+		reportRestartKey: restart,
+	}
+	if len(errors) > 0 {
+		report[reportErrorsKey] = errors
+	}
+	return report
+}
+
+// WatchController returns entity delta events for all models in the controller.
+func (w *Worker) WatchController() multiwatcher.Watcher {
+	return w.newWatcher(nil)
+}
+
+// WatchModel returns entity delta events just for the specified model.
+func (w *Worker) WatchModel(modelUUID string) multiwatcher.Watcher {
+	return w.newWatcher(
+		func(in []multiwatcher.Delta) []multiwatcher.Delta {
+			// Returns an empty slice if there is nothing to match with the
+			// implementation of the Watcher for noChanges. Both could potentially
+			// be updated to return a nil slice.
+			result := make([]multiwatcher.Delta, 0, len(in))
+			for _, delta := range in {
+				if delta.Entity.EntityID().ModelUUID == modelUUID {
+					result = append(result, delta)
+				}
+			}
+			return result
+		})
+}
+
+func (w *Worker) newWatcher(filter func([]multiwatcher.Delta) []multiwatcher.Delta) *Watcher {
+	watcher := &Watcher{
+		request: w.request,
+		control: &w.tomb,
+		logger:  w.config.Logger,
+		// Buffered err channel as if there is a fetch error on the all watcher backing
+		// the error is passed to the watcher.
+		err:    make(chan error, 1),
+		filter: filter,
+	}
+	w.mu.Lock()
+	w.watchers = append(w.watchers, watcher)
+	w.mu.Unlock()
+	return watcher
+}
+
+func (w *Worker) loop() error {
+	w.config.Logger.Tracef("worker loop started")
+	defer w.config.Logger.Tracef("worker loop completed")
+	defer func() {
+		if w.config.Cleanup != nil {
+			w.config.Cleanup()
+		}
+	}()
+
+	_ = w.config.PrometheusRegisterer.Register(w.metrics)
+	defer w.config.PrometheusRegisterer.Unregister(w.metrics)
+
+	for {
+		err := w.inner()
+		select {
+		case <-w.tomb.Dying():
+			return nil
+		default:
+			w.mu.Lock()
+			w.restartCount++
+			w.errors = append(w.errors, err)
+			if len(w.errors) > 5 {
```

Comment:
Magic number 5?

Reply:
Yeah. I didn't want to remember too many, but it wanted more than just the last one. Five seemed like a reasonable number. This bit is new. We could define a constant, but it is only used in one place. Perhaps a comment would be enough to explain?

Reply:
Quick comment would be fine I think.